File: sb-time.py
from collections import namedtuple
import time

seasons = (
    "Winter",
    "Spring",
    "Summer",
    "Autumn",
)

season_parts = (
    "Early",
    "Mid",
    "Late",
)

SkyBlockDate = namedtuple("SkyBlockDate", ("minute", "hour", "day", "month", "year"))

def get_skyblock_date(seconds: float) -> SkyBlockDate:
    year = int(seconds / 446_400) + 1
    # season_num = int(seconds / 111_600) % 4
    # season = seasons[season_num]
    month = int(seconds / 37_200 + 2) % 12 + 1
    day = int(seconds / 1200) % 31 + 1
    hour = int(seconds / 50) % 24
    minute = int(seconds % 50 * 60 / 50)
    return SkyBlockDate(minute, hour, day, month, year)

def get_num_suffix(n: int) -> str:
    if n // 10 == 1:
        return "th"
    if n % 10 == 1:
        return "st"
    elif n % 10 == 2:
        return "nd"
    elif n % 10 == 3:
        return "rd"
    return "th"

def get_season_name(month: int) -> str:
    return f"{season_parts[month % 3]} {seasons[month // 3 % 4]}"

def build_date_str(date: SkyBlockDate) -> str:
    day_suffix = get_num_suffix(date.day)
    season = get_season_name(date.month)
    return "{hour:02d}:{minute:02d}, {day}{day_suffix} of {season}({month}), year {year}".format(day_suffix=day_suffix, season=season, **date._asdict())

epoch_start = 1_560_275_700
# d = get_skyblock_date(0)
# print(build_date_str(d))
prev_msg_len = 0
while True:
    now = time.time()
    skyblock_sec = now - epoch_start
    skyblock_date = get_skyblock_date(skyblock_sec)
    msg = build_date_str(skyblock_date)
    print(f"\r{msg:<{prev_msg_len}}", end='')
    prev_msg_len = len(msg)
    try:
        time.sleep(50 / 60 - now % (50 / 60))
    except KeyboardInterrupt:
        print()
        exit()

File: sb-time.c
#include <math.h>
#include <stdio.h>
#include <time.h>

#define EPOCH_START 1560275700

char *seasons[4] = {
    "Winter",
    "Spring",
    "Summer",
    "Autumn",
};

char *season_parts[3] = {
    "Early",
    "Mid",
    "Late",
};

struct SkyBlockDate {
    unsigned int year:12;
    unsigned int month:4;
    unsigned int day:5;
    unsigned int hour:5;
    unsigned int minute:6;
};

void set_skyblock_date(double seconds, struct SkyBlockDate *data);
void set_sleep_time(double seconds, struct timespec *tv);
double tv2double(struct timespec *tv);
const char *get_num_suffix(unsigned int n);
void set_season_name(unsigned int month, char *season_name);
int set_msg(struct SkyBlockDate *data, char *msg);

int main()
{
    struct timespec tv;
    char msg[50];
    int prev_len = 0;
    do
    {
        struct SkyBlockDate data;

        clock_gettime(CLOCK_REALTIME, &tv);
        tv.tv_sec -= EPOCH_START;
        double now = tv2double(&tv);
        set_skyblock_date(now, &data);
        int msg_len = set_msg(&data, msg);

        printf("\r%-*s", prev_len, msg);
        for (int i = prev_len; i > msg_len; i--) printf("\b");
        fflush(stdout);

        prev_len = msg_len;

        set_sleep_time(now, &tv);
    }
    while(nanosleep(&tv, NULL) != -1);
    return 0;
}

void set_skyblock_date(double seconds, struct SkyBlockDate *data)
{
    data->year = (unsigned int) seconds / 446400 + 1;
    data->month = ((unsigned int) seconds / 37200 + 2) % 12 + 1;
    data->day = (unsigned int) seconds / 1200 % 31 + 1;
    data->hour = (unsigned int) seconds / 50 % 24;
    data->minute = (unsigned int) (fmod(seconds, 50) * 60) / 50;
}

void set_sleep_time(double seconds, struct timespec *tv)
{
    tv->tv_sec = 0;
    tv->tv_nsec = (long) ((50. / 60 - fmod(seconds, 50. / 60)) * 1e9);
}

double tv2double(struct timespec *tv)
{
    return (double) tv->tv_sec + (double) tv->tv_nsec * 1e-9;
}

const char *get_num_suffix(unsigned int n)
{
    if (n / 10 == 1)
        return "th";
    switch (n % 10)
    {
        case 1:
            return "st";
        case 2:
            return "nd";
        case 3:
            return "rd";
        default:
            return "th";
    }
}

void set_season_name(unsigned int month, char *season_name)
{
    char *part = season_parts[month % 3];
    char *season = seasons[month / 3 % 4];
    sprintf(season_name, "%s %s", part, season);
}

int set_msg(struct SkyBlockDate *data, char *msg)
{
    const char *day_suffix = get_num_suffix(data->day);
    char season_name[12];
    set_season_name(data->month, season_name);
    int len = snprintf(msg, 50, "%02u:%02u, %u%.2s of %s(%u), Year %u", data->hour, data->minute, data->day, day_suffix, season_name, data->month, data->year);
    return len;
}
